pkg/kubelet/util/manager/watch_based_manager.go

```diff
@@ -44,25 +45,103 @@ type isImmutableFunc func(runtime.Object) bool
 // objectCacheItem is a single item stored in objectCache.
 type objectCacheItem struct {
 	refCount  int
-	store     cache.Store
+	store     *cacheStore
+	reflector *cache.Reflector
+
 	hasSynced func() (bool, error)
 
-	// lock is protecting from closing stopCh multiple times.
-	lock   sync.Mutex
-	stopCh chan struct{}
+	// lock is to ensure the access and modify of lastAccessTime, stopped, and immutable are thread safety,
+	// and protecting from closing stopCh multiple times.
+	lock           sync.Mutex
+	lastAccessTime time.Time
+	stopped        bool
+	immutable      bool
+	stopCh         chan struct{}
 }
 
 func (i *objectCacheItem) stop() bool {
 	i.lock.Lock()
 	defer i.lock.Unlock()
-	select {
-	case <-i.stopCh:
-		// This means that channel is already closed.
+	return i.stopThreadUnsafe()
+}
+
+func (i *objectCacheItem) stopThreadUnsafe() bool {
+	if i.stopped {
 		return false
-	default:
-		close(i.stopCh)
-		return true
 	}
+	i.stopped = true
+	close(i.stopCh)
+	if !i.immutable {
+		i.store.unsetInitialized()
+	}
+	return true
+}
+
+func (i *objectCacheItem) setLastAccessTime(time time.Time) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	i.lastAccessTime = time
+}
+
+func (i *objectCacheItem) setImmutableIfUnset() {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	i.immutable = true
+}
+
+func (i *objectCacheItem) restartRefactor() {
+	stopCh := make(chan struct{})
+	i.stopCh = stopCh
+	i.stopped = false
+	go i.reflector.Run(stopCh)
```

Comment:
The problem is that we're running `reflector.Run` in its own gorotuine and it can happen than more than one is running at a time, which would be very bad.

Instead of that, we need to guarantee that we won't start `reflector.Run` until the previous one exited.
OTOH, `stopIfIdle` shouldn't block on that, because it is happening under the `objectCache` lock and may cause contention.

So I think the easiest way to solve it would be to:
- add sync.WaitGroup field to objectCacheItem
- switch both `reflector.Run()` calls to go func() { defer wg.Done(); reflector.Run() }
- before this call add `wg.Wait(); wg.Add(1)`
- in the original reflector.Run() add wg.Add(1)